velox/vector/VectorSaver.cpp

```diff
@@ -0,0 +1,650 @@
+    auto value = rawValues[i];
+    if (!value.isInline()) {
+      auto offset = *reinterpret_cast<const int64_t*>(
+          rawBytes + i * sizeof(StringView) + 8);
+      rawValues[i] =
+          StringView(computeStringPointer(offset, stringBuffers), value.size());
+    }
+  }
+}
+
+bool isVarcharOrVarbinary(const BaseVector& vector) {
+  return vector.typeKind() == TypeKind::VARCHAR ||
+      vector.typeKind() == TypeKind::VARBINARY;
+}
+
+void writeFlatVector(const BaseVector& vector, std::ostream& out) {
+  // Nulls buffer.
+  writeOptionalBuffer(vector.nulls(), out);
+
+  // Values buffer.
+  if (isVarcharOrVarbinary(vector)) {
+    const auto& values = vector.values();
+    if (values) {
+      write<int8_t>(1, out);
+
+      const auto& stringBuffers =
+          vector.asFlatVector<StringView>()->stringBuffers();
+      writeStringViews(vector.size(), values, stringBuffers, out);
+    } else {
+      write<int8_t>(0, out);
+    }
+  } else {
+    writeOptionalBuffer(vector.values(), out);
+  }
+
+  // String buffers.
+  if (isVarcharOrVarbinary(vector)) {
+    const auto& stringBuffers =
+        vector.asFlatVector<StringView>()->stringBuffers();
+    write<int32_t>(stringBuffers.size(), out);
+
+    for (const auto& buffer : stringBuffers) {
+      writeBuffer(buffer, out);
+    }
+  }
+}
+
+VectorPtr readFlatVector(
+    const TypePtr& type,
+    vector_size_t size,
+    std::istream& in,
+    memory::MemoryPool* pool) {
+  // Nulls buffer.
+  BufferPtr nulls = readOptionalBuffer(in, pool);
+
+  // Values buffer.
+  BufferPtr values = readOptionalBuffer(in, pool);
+
+  // String buffers.
+  std::vector<BufferPtr> stringBuffers;
+  if (type->isVarchar() || type->isVarbinary()) {
+    int32_t numStringBuffers = read<int32_t>(in);
+    for (auto i = 0; i < numStringBuffers; ++i) {
+      stringBuffers.push_back(readBuffer(in, pool));
+    }
+
+    // Update the pointers in the StringViews.
+    if (values) {
+      restoreVectorStringViews(size, values, stringBuffers);
+    }
+  }
+
+  return VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH_ALL(
+      createFlat, type->kind(), type, size, pool, nulls, values, stringBuffers);
+}
+
+template <TypeKind kind>
+void writeConstant(const BaseVector& vector, std::ostream& out) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  auto value = vector.as<ConstantVector<T>>()->valueAt(0);
+  out.write((const char*)&value, sizeof(T));
+
+  if constexpr (std::is_same_v<T, StringView>) {
+    if (!value.isInline()) {
+      write<int32_t>(value.size(), out);
+      out.write(value.data(), value.size());
+    }
+  }
+}
+
+void writeConstantVector(const BaseVector& vector, std::ostream& out) {
+  bool isNull = vector.isNullAt(0);
+  write<bool>(isNull, out);
+
+  if (isNull) {
+    return;
+  }
+
+  auto baseVector = vector.valueVector();
+  write<bool>(baseVector == nullptr, out);
+
+  if (baseVector) {
+    saveVector(*baseVector, out);
+    write<int32_t>(vector.wrappedIndex(0), out);
```

Comment:
valueVector returns the vector after removing only the constant wrapping however wrappedIndex() returns the index after removing all wrappings. Should we use index() here instead?

Reply:
That's a good point. Technically speaking we should use index(). In practice though, constant vector cannot wrap neither constant nor dictionary, hence, index() will be same as wrappedIndex(). To call index(), I'll need to add a type dispatch to cast to ConstantVector<T>, since index() method is not available through BaseVector. 

```
    VELOX_CHECK_NE(
        base->BaseVector::encoding(),
        VectorEncoding::Simple::CONSTANT,
        "Constant vector cannot wrap Constant vector");
    VELOX_CHECK_NE(
        base->BaseVector::encoding(),
        VectorEncoding::Simple::DICTIONARY,
        "Constant vector cannot wrap Dictionary vector");
```